src/vs/workbench/contrib/terminal/browser/terminalGroupService.ts

```diff
@@ -0,0 +1,316 @@
+import { TerminalGroup } from 'vs/workbench/contrib/terminal/browser/terminalGroup';
+import { KEYBINDING_CONTEXT_TERMINAL_GROUP_COUNT } from 'vs/workbench/contrib/terminal/common/terminal';
+
+export class TerminalGroupService extends Disposable implements ITerminalGroupService {
+	declare _serviceBrand: undefined;
+
+	groups: ITerminalGroup[] = [];
+	activeGroupIndex: number = -1;
+	get instances(): ITerminalInstance[] {
+		return this.groups.reduce((p, c) => p.concat(c.terminalInstances), [] as ITerminalInstance[]);
+	}
+	activeInstanceIndex: number = -1;
+
+	private _terminalGroupCountContextKey: IContextKey<number>;
+	private _container: HTMLElement | undefined;
+
+	private readonly _onDidChangeActiveGroup = new Emitter<ITerminalGroup | undefined>();
+	get onDidChangeActiveGroup(): Event<ITerminalGroup | undefined> { return this._onDidChangeActiveGroup.event; }
+	private readonly _onDidDisposeGroup = new Emitter<ITerminalGroup>();
+	get onDidDisposeGroup(): Event<ITerminalGroup> { return this._onDidDisposeGroup.event; }
+	private readonly _onDidChangeGroups = new Emitter<void>();
+	get onDidChangeGroups(): Event<void> { return this._onDidChangeGroups.event; }
+
+	private readonly _onDidDisposeInstance = new Emitter<ITerminalInstance>();
+	get onDidDisposeInstance(): Event<ITerminalInstance> { return this._onDidDisposeInstance.event; }
+	private readonly _onDidChangeActiveInstance = new Emitter<ITerminalInstance | undefined>();
+	get onDidChangeActiveInstance(): Event<ITerminalInstance | undefined> { return this._onDidChangeActiveInstance.event; }
+	private readonly _onDidChangeInstances = new Emitter<void>();
+	get onDidChangeInstances(): Event<void> { return this._onDidChangeInstances.event; }
+
+	private readonly _onPanelOrientationChanged = new Emitter<Orientation>();
+	get onPanelOrientationChanged(): Event<Orientation> { return this._onPanelOrientationChanged.event; }
+
+	constructor(
+		@IContextKeyService private _contextKeyService: IContextKeyService,
+		@IInstantiationService private readonly _instantiationService: IInstantiationService
+	) {
+		super();
+
+		this.onDidDisposeGroup(group => this._removeGroup(group));
+
+		this._terminalGroupCountContextKey = KEYBINDING_CONTEXT_TERMINAL_GROUP_COUNT.bindTo(this._contextKeyService);
+		this.onDidChangeGroups(() => this._terminalGroupCountContextKey.set(this.groups.length));
+	}
+
+	get activeGroup(): ITerminalGroup | undefined {
+		if (this.activeGroupIndex < 0 || this.activeGroupIndex >= this.groups.length) {
+			return undefined;
+		}
+		return this.groups[this.activeGroupIndex];
+	}
+	set activeGroup(value: ITerminalGroup | undefined) {
+		if (value === undefined) {
+			this.activeGroupIndex = -1;
+			return;
+		}
+		this.activeGroupIndex = this.groups.findIndex(e => e === value);
+	}
+
+	get activeInstance(): ITerminalInstance | undefined {
+		return this.activeGroup?.activeInstance;
+	}
+
+	set activeInstance(value: ITerminalInstance | undefined) {
+		if (value === undefined) {
```

Comment:
will we ever want to set activeInstance to undefined? for instance when there are 0 terminals?

Reply:
Good question as it isn't very clear from the code. The main reason for that is because I need to align the setter types with the getter types, we actually only want to allow setting to `ITerminalInstance` externally. Whether `activeInstance` is set to `undefined` should only be controlled by the group service or the editor service. I'll change this back to a `setActiveInstance` method to avoid this compile time issue slipping though.